enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,260 @@
+
+## Motivation
+
+In many cases the default out of the box SCCs are sufficient as they cover the most
+common patterns for controlling permissions pods can have.
+Additionally, the most recent enablement of Pod Security Admission also covers
+an additional layer of protection next to OpenShift SCCs.
+
+However, OpensShift historically does allow cluster administrators to add
+additional custom SCCs on top of the default out of the box SCCs.
+Custom SCCs are not considered different from the perspective of
+the SCC admission plugin.
+
+When a pod is being admitted, the admission plugin lists all SCCs available in the system, including all custom SCCs.
+The list of SCCs is then sorted by:
+- SCC priority in descending order
+- SCCs with the same priority: SCC restriction score in descending order (most restrictive first, least restrictive last)
+- SCCs with the same restriction score: lexical ordering of the SCC name in descending order
+
+An SCC is considered applicable if the requesting user or the
+pod's service account has `use` permissions against the `security.openshift.io.securitycontextconstraints` resource
+for an SCC in the requested namespace.
+If an SCC is applicable, the pod's manifest is potentially mutated according to the SCC logic.
+
+One can see that custom SCCs can easily overrule existing out of the box SCCs
+if the pod's service account or the requesting user has sufficient privileges to use any of them.
+
+In that case, if a custom SCC has higher priority, is more restrictive, or it's name is lexically ordered before an out of the box SCC,
+any OpenShift out of the box core pod will be mutated according to the custom SCC specified behavior.
+
+This can lead to unknown mutations of core OpenShift workloads.
+Worst case it can render OpenShift core workloads dysfunctional
+if i.e. a custom SCC is more restrictive than a previously applicable less restrictive out of the box SCC that was applicable to that workload.
+
+This proposal includes two suggested protection modes, a preferred "hard protection" and a "soft protection" variant
+which are described below.
+
+We need a mechanism to protect OpenShift core workloads such that they are not
+suspect of pod mutation due to user provided custom SCCs.
+
+### User Stories
+
+1. As an OpenShift core developer I want to make sure that my shipped workload
+is guaranteed to be associated with an out of the box SCC.
+
+2. As an OpenShift core developer I want to make sure that my shipped workload
+is not preempted by custom user provided SCCs.
+
+### Goals
+
+1. Protect core OpenShift workloads from being preempted by user provided custom SCCs.
+
+2. Be able to pin core OpenShift workloads to a concrete out of the box SCC.
+
+### Non-Goals
+
+1. The proposed solution is not supposed to be an official API to be used by end users.
+The mechanism to pin to SCCs is supposed to be used by core OpenShift workloads only
+to protect workloads shipped by the product.
```

Comment:
I believe instead of just claiming it for internal purpose, we could submit it to the graduation process.

We would need to version the annotation and specifically say that we will absolutely certainly break alpha support and won't care about any bugs this may cause. Also, I would not document it at that point.

I don't want to end up with alpha/beta annotations like with the service-ca ones.

Reply:
that seems reasonable, though i suspect once you introduce the alpha annotation for internal use, it *is* going to get discovered and used by customers/support to solve SCC problems.  My guess is you'll end up leaving code  for handling the alpha/beta annotations in place, but having them behave identically to whatever you call the GA ones when you get there.

one potential reason @eparis raised for why exposing for use by non-payload workloads could be undesirable is that it constrains our ability to introduce new SCCs in the future (as was done for restricted-v2) since we have less assurance that all workloads are actually using the SCC we might expect/will pick up the new SCC.

but arguably that problem already exists today as seen w/ the recent logging bug(where a custom SCC added to the cluster got picked up by a workload in preference to the default SCCs), so maybe that concern is irrelevant.

Reply:
> it is going to get discovered and used by customers/support to solve SCC problems.

We cannot prevent misbehavior or unsupported behavior in all cases. At the end, being cluster admin, users can do anything. We just don't want to make it easy.

Also, we _could_ decide to go with the "soft" variant in this OEP for GA use for end users, as the currently favored variant will fail admission if your workload does not have permission to use the required SCC. In this case, behavior would indeed be different.

Reply:
But generally, I am +1 on making this public. Marking this as non-public was a protective measure to prevent side-effects we didn't see but I definitely see the benefit of this annotation for a wide audience 👍

Reply:
> We cannot prevent misbehavior or unsupported behavior in all cases. At the end, being cluster admin, users can do anything. We just don't want to make it easy.

agreed, i just wanted to call it out

> Also, we could decide to go with the "soft" variant in this OEP for GA use for end users

i think the hard variant is better....the current challenge (welll one of them) w/ SCCs is the indeterminate behavior and it not being obvious what you got or why. So providing another mechanism that can still have that result isn't as helpful as providing a mechanism where it's "what you asked for or nothing"

Reply:
marked as generally available for end users.